Sources/Subjects/ReplaySubject.swift

```diff
@@ -37,53 +37,51 @@ public final class ReplaySubject<Output, Failure: Error>: Subject {
     }
 
     public func send(_ value: Output) {
-        lock.lock()
-        defer { lock.unlock() }
-
         guard isActive else { return }
 
+        lock.lock()
         buffer.append(value)
+        lock.unlock()
 
         if buffer.count > bufferSize {
+            lock.lock()
             buffer.removeFirst()
+            lock.unlock()
         }
 
         subscriptions.forEach { $0.forwardValueToBuffer(value) }
     }
```

Comment:
More fine-grained locking is moving in the right direction, but unfortunately this introduces some other problems:

- `isActive` is no longer protected by the lock, so accesses could race. This might result in thread 1 seeing `isActive == true` while thread 2 is trying to complete or cancel the subject.
- `buffer.append()` is locked, but then `buffer.count` is accessed immediately afterward. This could crash if thread 1 sees a count of 1, while thread 2 is removing the last item from the buffer, then calls `buffer.removeFirst()` on an empty buffer.
- Thread 2 could change `subscriptions` while thread 1 is sending the new value, resulting in an off-by-one error or subscribers missing events.

I'd recommend a pattern like this:

```swift
public func send(_ value: Output) {
  let subscriptions: [Subscription<AnySubscriber<Output, Failure>>]

  do { // critical section
    lock.lock()
    defer { lock.unlock() }

    guard isActive else { return }

    buffer.append(value)
    if buffer.count > bufferSize {
      buffer.removeFirst()
    }

    subscriptions = self.subscriptions
  }

  subscriptions.forEach { $0.forwardValueToBuffer(value) }
}
```

Reply:
Updated! Followed suit for the other methods and I thiiink I got it right — appreciate the review pass here. 🙏🏽